### DISCUSSION TITLE
Compiling translations while deploying to Heroku

### DISCUSSION THREAD

**Discussion by davidwessman** [fnando/i18n-js#710] [Q&A]
Hello!

Thank you for a great gem, we are working on migrating from v3 to v4 and the workflow works great for development.
Right now we are committing the file with translations to git, but if we could get it to work on deploy we could gitignore the file and consider it an asset instead.

Do anyone have a nice way for configuring it on Heroku Deploy?

We already have some hooks into the `assets:precompile` to cleanup unused assets, but since there is no longer a rake-task  that is not as straight forward.

**Comment by sheeehy** [comment]
you can write a custom task that combines the generation of i18n-js translations and any other operations you perform during assets:precompile.

Create a file named lib/tasks/custom_i18n_js.rake:
```
namespace :i18n do
  desc "Export translations to JS"
  task :export => :environment do
    # You can invoke i18n-js export here
    I18n::JS.export
  end
end

Rake::Task["assets:precompile"].enhance do
  Rake::Task["i18n:export"].invoke
end
```

then add the file to gitignore and execute heroku build,

**Reply by davidwessman** [reply]
Thank you for a great start of the script!
I had to hook into an earlier task to make sure that my translations are prepared when assets are compiled by Shakapacker.

```ruby

  namespace :i18n do
    desc("Export translations to JS")
    task(export: :environment) do
      puts("Exporting translations to JS")
      I18nJS.call(config_file: "config/i18n.yml")
    end
  end

# Need to enhance `shakapacker:verify_install` instead of `assets:precompile`
# to make sure that our translations are exported shakapacker compiles.
Rake::Task["shakapacker:verify_install"].enhance do
  Rake::Task["i18n:export"].invoke
end
```

**Comment by koppen** [comment]
We recently upgraded an app to v4 and ran into problems trying to deploy to Heroku as well. And while the above rake tasks provided excellent pointers, we needed a bit more trickery to make things work, so I figured I might as well leave the tale here.

First, a little bit about our setup:

* i18n-js 4
* rails 6
* jsbundling 1.2
  * Using webpack 5.9
* Deployed to Heroku running nodejs- and the ruby buildpacks

When we tried to deploy we got the following error during Herokus build phase:

```
ERROR in ./app/javascript/i18n.js 3:0-47
       Module not found: Error: Can't resolve './translations.json' in '/tmp/build_36012f77/app/javascript'
       resolve './translations.json' in '/tmp/build_36012f77/app/javascript'
```

which indicates the translations.json file isn't getting exported, even though we had a rake task like the above:

```ruby
namespace :i18n do
  namespace :js do
    desc "Exports translations to be used by i18n-js"
    task :export => :environment do
      `bundle exec i18n export`
    end
  end
end

# Run `i18n export` prior to building Javascript assets, so translations are
# available for use. javascript:build is the rake task run by jsbundling
# during assets:precompile
Rake::Task["javascript:build"].enhance(["i18n:js:export"])
```

Turns out we were running the nodejs buildpack first (since that's [the recommended order](https://devcenter.heroku.com/articles/ruby-support#node-js-support) in order to control Node and Yarn versions). The nodejs buildpack by default [runs your `build` script](https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process) if you have one defined in `package.json`. `jsbundling` adds a `build` script to `package.json`, which it uses internally, I guess:

```json
  "scripts": {
    "build": "webpack --config webpack.config.js"
  }
```

Unfortunately, this meant that the nodejs buildpack would try to compile our assets before the above rake task was even invoked and even before our Ruby dependencies had been installed. There's no chance for our translations to have been exported at this point.

Luckily Heroku provides a way to [customize the build process](https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process), which we can use to tell Heroku to not build anything during the build step of the nodejs buildpack:

```json
  "scripts": {
    "build": "webpack --config webpack.config.js",
    "heroku-postbuild": ""
  }
```

With this in place in `package.json` Heroku would do nothing during the nodejs build step and everything would work just fine when we reached the ruby build step:

```
-----> Build
       Detected both "build" and "heroku-postbuild" scripts
       Running heroku-postbuild (yarn)

-----> Pruning devDependencies
...
-----> Preparing app for Rails asset pipeline
       Running: rake assets:precompile
```

Here's hoping this saves someone else the headaches I had figuring it out.
